src/transformers/models/patchtst/modeling_patchtst.py

```diff
@@ -0,0 +1,1958 @@
+        self.use_cls_token = config.use_cls_token
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        x: tensor [bs x nvars x num_patches x d_model]
+                or [bs x nvars x (num_patches+1) x d_model] if use cls_token
+        output: tensor [bs x nvars x num_patches x patch_length]
+        """
+        x = self.linear(self.dropout(x))  # [bs x nvars x num_patches x patch_length]
+        if self.use_cls_token:
+            x = x[:, :, 1:, :]  # remove the first cls token
+        return x
+
+
+class PatchTSTForMaskPretraining(PatchTSTPreTrainedModel):
+    # PatchTSTModel + Pretraining Head
+    def __init__(self, config: PatchTSTConfig):
+        super().__init__(config)
+
+        config.mask_input = True
+        self.model = PatchTSTModel(config=config)
+        self.head = MaskPretrainHead(config)
+        self.loss = torch.nn.MSELoss(reduction="none")
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def forward(
+        self,
+        past_values: torch.Tensor,
+        past_observed_mask: Optional[torch.Tensor] = None,
+        future_values: Optional[torch.Tensor] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, PatchTSTForMaskPretrainingOutput]:
+        """
+        past_values (x): tensor [bs x sequence_length x num_input_channels ] future_values (y): labels
+        """
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        # past_values: [bs x nvars x num_patches x d_model] or
+        # [bs x nvars x (num_patches+1) x d_model] if use cls_token
+        model_output = self.model(past_values, output_hidden_states=output_hidden_states)
+
+        # model_output[0]: [bs x nvars x num_patches x patch_length] or
+        # [bs x nvars x (num_patches+1) x patch_length] if use cls_token
+        x_hat = self.head(model_output[0])
+
+        # calculate masked_loss
+        loss_val = self.loss(x_hat, model_output.patched_input)
+        masked_loss = (loss_val.mean(dim=-1) * model_output.mask).sum() / (model_output.mask.sum() + 1e-10)
+
+        encoder_states = model_output.hidden_states
+        if not return_dict:
+            return tuple(v for v in [masked_loss, x_hat, encoder_states] if v is not None)
+        return PatchTSTForMaskPretrainingOutput(
+            loss=masked_loss, prediction_output=x_hat, hidden_states=encoder_states
+        )
+
+
+class PatchTSTForClassification(PatchTSTPreTrainedModel):
+    # PatchTST model + classification head
+    def __init__(self, config: PatchTSTConfig):
+        super().__init__(config)
+
+        self.model = PatchTSTModel(config)
+        self.head = ClassificationHead(config)
+        self.loss = nn.CrossEntropyLoss()
```

Comment:
we usually only define the loss in the forward

Reply:
Would be great if we could define the loss only in the forward in case the model has `labels` defined